### PR TITLE
Fix low charge error

### PR DESCRIPTION
**Pull request by jonflynng**
We have seen an error in the product-move-api caused because the amount a user would have to pay to switch from a recurring contribution to supporter plus is less than the minimum amount Stripe will allow as a charge. This is $0.50 or equivalent in other currencies.

https://docs.google.com/document/d/1OyOraR45-W7c_QPkgP2bJW_QT0nwnqFZHMeeAba6jak/edit#

### CODE REVIEW

File: handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/move/ProductMoveEndpointTypes.scala [hunk 1]
```diff
@@ -20,6 +20,7 @@ object ProductMoveEndpointTypes {
   case class ExpectedInput(
       @description("Price of new Supporter Plus subscription") price: BigDecimal,
       @description("Whether to preview the move or to carry it out") preview: Boolean,
+      recallPreview: Boolean,
```
Comment (jonflynng) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/move/ProductMoveEndpointTypes.scala:23]: probably want a better name for this

File: handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala [hunk 2]
```diff
@@ -282,115 +292,164 @@ object RecurringContributionToSupporterPlus {
...
+            EmailPayload(
+              Address = Some(account.billToContact.workEmail),
+              ContactAttributes = EmailPayloadContactAttributes(
+                SubscriberAttributes = EmailPayloadProductSwitchAttributes(
+                  first_name = account.billToContact.firstName,
+                  last_name = account.billToContact.lastName,
+                  currency = account.basicInfo.currency.symbol,
+                  price = price.setScale(2, BigDecimal.RoundingMode.FLOOR).toString,
+                  first_payment_amount =
+                    if (collectPayment) paidAmount.setScale(2, BigDecimal.RoundingMode.FLOOR).toString else "0.00",
```
Comment (jonflynng) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala:396]: These if statements are used throughout, not sure if this is the best way to do it.

File: handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/move/ProductMoveEndpointTypes.scala (hunk 1 above)
Comment (rupertbates) [reply handlers/product-move-api/src/main/scala/com/gu/productmove/endpoint/move/ProductMoveEndpointTypes.scala:23]: It needs a description  as well. Can we call it something like `checkChargeAmountBeforeUpdate`? It's verbose but does describe what needs to happen

File: handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala [hunk 3]
```diff
@@ -282,115 +292,164 @@ object RecurringContributionToSupporterPlus {
...
+              subscriptionName,
+              price,
+              billingPeriod,
+              ratePlanCharge,
+              currency,
+              currentRatePlan.id,
+            )
+            amount = previewResponse.asInstanceOf[PreviewResult].amountPayableToday
+          } yield amount
+        } else ZIO.succeed(BigDecimal(1))
```
[the remark this answers is not in the document]
Comment (rupertbates) [reply handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala:330]: I would refactor this into a separate method called `shouldCollectPayment` which returns a boolean and includes the conditional below

File: handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala [hunk 4]
```diff
@@ -282,115 +292,164 @@ object RecurringContributionToSupporterPlus {
...
+
+      collectPayment = !(amountPayableToday > BigDecimal(0) && amountPayableToday < BigDecimal(0.50))
```
Comment (rupertbates) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala:332]: Will we have the same issue with negative amounts? Eg. -$0.40

File: handlers/product-move-api/src/main/scala/com/gu/productmove/zuora/GetInvoiceItems.scala
Comment (rupertbates) [inline handlers/product-move-api/src/main/scala/com/gu/productmove/zuora/GetInvoiceItems.scala:1]: There are already classes for this and `InvoiceItemAdjustment` in the project, is it not possible to reuse those?

File: handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala (hunk 2 above)
Comment (rupertbates) [reply handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala:396]: I think they're fine

File: handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala [hunk 5]
```diff
@@ -282,115 +292,164 @@ object RecurringContributionToSupporterPlus {
...
-                price = price.setScale(2, BigDecimal.RoundingMode.FLOOR).toString,
-                first_payment_amount = paidAmount.setScale(2, BigDecimal.RoundingMode.FLOOR).toString,
-                date_of_first_payment = todaysDate.format(DateTimeFormatter.ofPattern("d MMMM uuuu")),
-                payment_frequency = billingPeriodValue + "ly",
-                subscription_id = subscriptionName.value,
+      supporterPlusRatePlanIds <- ZIO.fromEither(getSupporterPlusRatePlanIds(stage, billingPeriod))
+
+      _ <-
+        if (!collectPayment) {
+          for {
```
[the remark this answers is not in the document]
Comment (jonflynng) [reply handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala:354]: Yep, will do both.

File: handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonflynng) [reply handlers/product-move-api/src/main/scala/com/gu/productmove/switchtype/RecurringContributionToSupporterPlus.scala:330]: Yes, and if `false`, the `amountPayableToday` will default to 0.

**Pull request merged**